### ISSUE TITLE
Can't train samediff model when reloaded

### ISSUE THREAD

**Issue by SidneyLann** [deeplearning4j/deeplearning4j#9145]
#### Issue Description

Please describe our issue, along with:
- expected behavior
Continue to train
- encountered behavior
java.lang.NullPointerException
in 
sd = CheckpointListener.loadLastCheckpoint(modelFolder, true);

#### Version Information
Please indicate relevant versions, including, if relevant:

* Deeplearning4j version beta7
* Platform information (OS, etc) win7
* CUDA version, if used no
* NVIDIA driver version, if in use no

#### Additional Information

Where applicable, please also provide:

* Full log or exception stack trace (ideally in a Gist: gist.github.com)
* pom.xml file or similar (also in a Gist)

java.lang.NullPointerException
	at org.nd4j.linalg.api.ops.impl.layers.recurrent.LSTMLayerBp.<init>(LSTMLayerBp.java:54)
	at org.nd4j.linalg.api.ops.impl.layers.recurrent.LSTMLayer.doDiff(LSTMLayer.java:137)
	at org.nd4j.autodiff.functions.DifferentialFunction.diff(DifferentialFunction.java:559)
	at org.nd4j.autodiff.samediff.SameDiff$1.define(SameDiff.java:4427)
	at org.nd4j.autodiff.samediff.SameDiff.defineFunction(SameDiff.java:3971)
	at org.nd4j.autodiff.samediff.SameDiff.defineFunction(SameDiff.java:3956)
	at org.nd4j.autodiff.samediff.SameDiff.createGradFunction(SameDiff.java:4167)
	at org.nd4j.autodiff.samediff.SameDiff.createGradFunction(SameDiff.java:4074)
	at org.nd4j.autodiff.samediff.SameDiff.fitHelper(SameDiff.java:1647)
	at org.nd4j.autodiff.samediff.SameDiff.fit(SameDiff.java:1569)
	at org.nd4j.autodiff.samediff.SameDiff.fit(SameDiff.java:1509)
	at org.nd4j.autodiff.samediff.config.FitConfig.exec(FitConfig.java:173)

#### Contributing

If you'd like to help us fix the issue by contributing some code, but would
like guidance or help in doing so, please mention it!

**Comment by agibsonccc** [comment]
@SidneyLann  could you give me a sample model for this to test? We're finishing up onnx and TF import rewrite and will be looking at retraining to verify what's missing. Anything there you can do to help would be great. Thanks!

**Comment by SidneyLann** [comment]
1. Comment //throw new IllegalStateException("Detected existing checkpoint files at directory " + rootDir.getAbsolutePath() +...in CheckpointListener.java
2. Replace by below content in main() in CustomListenerExample.java:

		SameDiff sd;
		File modelFolder = new File("./model");

		if (modelFolder.list().length > 1) {
			System.out.println("Load model...");
			sd = CheckpointListener.loadLastCheckpoint(modelFolder, true);
		} else {
			System.out.println("Build model...");
			sd = makeMNISTNet();

			// Create and set the training configuration
			double learningRate = 1e-3;
			TrainingConfig config = new TrainingConfig.Builder().l2(1e-4) // L2 regularization
					.updater(new Adam(learningRate)) // Adam optimizer with specified learning rate
					.dataSetFeatureMapping("input") // DataSet features array should be associated with variable "input"
					.dataSetLabelMapping("label") // DataSet label array should be associated with variable "label
					.addEvaluations(false, "out", 0, new Evaluation()).build();

			sd.setTrainingConfig(config);
		}
		CheckpointListener checkpointListener = CheckpointListener.builder(modelFolder).saveEveryNEpochs(1)
				.deleteExisting(false).build();
		sd.setListeners(checkpointListener);

		int batchSize = 32;
		DataSetIterator trainData = new MnistDataSetIterator(batchSize, true, 12345);

		// Perform training
		History hist = sd.fit().train(trainData, 1).exec();
		Evaluation e = hist.finalTrainingEvaluations().evaluation("out");

		System.out.println("Accuracy: " + e.accuracy());

		CustomListener listener = new CustomListener();

		sd.output().data(new MnistDataSetIterator(10, 10, false, false, true, 12345)).output("out").listeners(listener)
				.exec();

		System.out.println("Z: " + listener.z);
		System.out.println("Out (softmax(z)): " + listener.out);
	
3. Run CustomListenerExample.java
4. Run CustomListenerExample.java after 3. finished(this step should reproduce this issue)

**Comment by SidneyLann** [comment]
[main] WARN org.nd4j.imports.converters.ImportClassMapping - Duplicate TF op mapping found for op Pow: org.nd4j.linalg.api.ops.impl.scalar.Pow vs org.nd4j.linalg.api.ops.impl.transforms.custom.Pow
[main] WARN org.nd4j.imports.converters.ImportClassMapping - Duplicate TF op mapping found for op FloorMod: org.nd4j.linalg.api.ops.impl.transforms.pairwise.arithmetic.FModOp vs org.nd4j.linalg.api.ops.impl.transforms.pairwise.arithmetic.FloorModOp
java.lang.NullPointerException
	at org.nd4j.linalg.api.ops.impl.layers.recurrent.LSTMLayer.configureWithSameDiff(LSTMLayer.java:210)
	at org.nd4j.autodiff.samediff.SameDiff.fromFlatBuffers(SameDiff.java:5503)
	at org.nd4j.autodiff.samediff.SameDiff.fromFlatFile(SameDiff.java:5359)
	at org.nd4j.autodiff.samediff.SameDiff.load(SameDiff.java:5250)
	at org.nd4j.autodiff.listeners.checkpoint.CheckpointListener.loadCheckpoint(CheckpointListener.java:407)
	at org.nd4j.autodiff.listeners.checkpoint.CheckpointListener.loadLastCheckpoint(CheckpointListener.java:417)



Can't reload model to train in SNAPSHOT

**Comment by agibsonccc** [comment]
@SidneyLann you still never gave me a model like I asked here at the beginning of the year. I won't spend time on issues where I'm not getting any collaboration. I will not dedicate my time trying to reverse engineer a model that could potentially reproduce this problem. Please give me a model or I'm going to focus my attention on issues I myself see or issues that have more information for them.

**Comment by SidneyLann** [comment]
The steps I commented on 1 Jan will generate the model, but this case has no problem now. the new case I commented on last post still has problem. I will DM you, give you code to generate model and then can use to reload. Thank you.

**Comment by agibsonccc** [comment]
@SidneyLann Appreciated. I was missing the makeModelMnist() you had there though. It wasn't clear how you were creating the model. I mainly saw the code for creating some sort of model and you loading it from a checkpoint.

**Comment by SidneyLann** [comment]
java.lang.NullPointerException: Cannot load from object array because "inputsForOp" is null
	at org.nd4j.linalg.api.ops.impl.layers.recurrent.LSTMLayer.configureWithSameDiff(LSTMLayer.java:210)
	at org.nd4j.autodiff.samediff.SameDiff.fromFlatBuffers(SameDiff.java:5572)
	at org.nd4j.autodiff.samediff.SameDiff.fromFlatFile(SameDiff.java:5428)
	at org.nd4j.autodiff.samediff.SameDiff.load(SameDiff.java:5319)
	at org.nd4j.autodiff.listeners.checkpoint.CheckpointListener.loadCheckpoint(CheckpointListener.java:407)
	at org.nd4j.autodiff.listeners.checkpoint.CheckpointListener.loadLastCheckpoint(CheckpointListener.java:417)




Still can't reload SameDiff model in a new case in M2 or SNAPSHOT

**Comment by agibsonccc** [comment]
@SidneyLann could you directly respond to what I asked you in December? You keep pinging me about this issue yet provide me *zero* code to actually reproduce this. I'm not going to spend time on something where I have to jump through hoops just to reproduce your issue. I need to be able to reproduce this. If you can meet me in the middle this will get solved quick.

I need you to be more cooperative though. Your bug reports are always half of what I need to make something work. 
I don't want a half done bug report. I want a class I can import in to my IDE and run. I don't think that's a lot to ask for.

Edit: Also you'll know something is fixed if I close an issue. Tasks I pick to work on are very deliberate. Tasks that are half filed don't get prioritized. So hoping for a magic fix after a release isn't really going to help.

**Comment by SidneyLann** [comment]
1. In fact, they are new subissues when I give new info, it is not the same error.
2. I had given you my project, models and data, because the converted bert model has more than 350M and can't send you by email, can you debug it on my desktop in remote if you are not remain my project files?

**Comment by agibsonccc** [comment]
@SidneyLann thanks for the quick response. I really just need a reproducer. Could you give me a similar model architecture? Anything trained on MNIST shouldn't be that big. I don't need the exact model as long as it's something to reproduce the issue. If it's just LSTMs acting up for example just send me that.

**Comment by SidneyLann** [comment]
Simple example may not reproduce this issue, let me try first then.

**Comment by agibsonccc** [comment]
@SidneyLann if it doesn't then that's a different story. Then just send me the model via drive and community DM like last time.

**Comment by SidneyLann** [comment]
Do you still have my bert_pretrain_model_zh.pb sent to you last time? I can't access google driver and it is sent to you by my qq friend who is in deeplearning4j qq group, but I can't find him now. If you still have my bert_pretrain_model_zh.pb, the project files and data are easy to send to you.

**Comment by SidneyLann** [comment]
https://drive.google.com/drive/folders/1VbwKHzRiqx45LbzJAUyaGpSIObfGVRHa?usp=sharing

**Comment by SidneyLann** [comment]
My bert_pretrain_model_zh.pb is here, so I just need to send you my project and data. Thanks.

**Comment by agibsonccc** [comment]
@SidneyLann I have your BERT model yes but that model loaded fine. This is still a different one. All I need is a samediff graph with the right declarations for it to break. It could be the way you're using it which is why I need to see how exactly you did it.

**Comment by SidneyLann** [comment]
Sent project files to you by email. Thanks.

**Comment by SidneyLann** [comment]
Hi, any findings?

**Comment by agibsonccc** [comment]
@SidneyLann fixed this. Configuring the samediff instance came before the inputs were available causing the nullpointer. I'll close it with this PR. Thanks for helping reproduce this.

**Comment by agibsonccc** [comment]
@SidneyLann it appears your other issue was due to using the old import. I am not supporting that. Please use the new one.

**Comment by SidneyLann** [comment]
I train a new model and restore to train, still has the problem:
Cannot load from object array because "inputsForOp" is null
	at org.nd4j.linalg.api.ops.impl.layers.recurrent.LSTMLayer.configureWithSameDiff(LSTMLayer.java:210)


The fix has not apply to SNAPSHOT?

**Comment by SidneyLann** [comment]
java.lang.NullPointerException: Cannot load from object array because "inputsForOp" is null
	at org.nd4j.linalg.api.ops.impl.layers.recurrent.LSTMLayer.configureWithSameDiff(LSTMLayer.java:210)
	at org.nd4j.autodiff.samediff.SameDiff.fromFlatBuffers(SameDiff.java:5503)
	at org.nd4j.autodiff.samediff.SameDiff.fromFlatFile(SameDiff.java:5359)
	at org.nd4j.autodiff.samediff.SameDiff.load(SameDiff.java:5250)
	at org.nd4j.autodiff.listeners.checkpoint.CheckpointListener.loadCheckpoint(CheckpointListener.java:407)
	at org.nd4j.autodiff.listeners.checkpoint.CheckpointListener.loadLastCheckpoint(CheckpointListener.java:417)



latest SNAPSHOT still has this problem.
